<a href="https://colab.research.google.com/github/rani-sikdar/Tensorflow-Learning/blob/main/Pytorch_basics_binary_classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import torch.nn as nn
import pandas as pd
import numpy as np
import torch.optim as optim
import torch

In [2]:
# generate the data

np.random.seed(42)   # same like random_state in ML ..... the X_data will be same everytime
X_data = np.random.uniform(low=1, high=100, size=100)
X_data

array([38.07947177, 95.12071633, 73.46740024, 60.26718994, 16.4458454 ,
       16.44345751,  6.7502776 , 86.75143843, 60.51038616, 71.0991852 ,
        3.03786494, 97.02107536, 83.41182144, 22.02157196, 19.00067175,
       19.15704648, 31.11998205, 52.95088673, 43.76255685, 29.83168488,
       61.57343658, 14.8098922 , 29.9223202 , 37.26982249, 46.15092844,
       78.73242018, 20.76770443, 51.9092094 , 59.64904232,  5.59859086,
       61.14694034, 17.88188825,  7.44010771, 94.93966819, 96.59757127,
       81.03133746, 31.15676315, 10.66953929, 68.73906962, 44.57509688,
       13.08178525, 50.0225141 ,  4.40446359, 91.02271981, 26.61921818,
       66.58970615, 31.85939653, 52.4867341 , 55.12431765, 19.3005911 ,
       96.98887815, 77.73814951, 94.01039521, 89.58790769, 60.1920979 ,
       92.26554927,  9.7607577 , 20.40230338,  5.4775016 , 33.20770275,
       39.47905168, 27.86355415, 83.04501341, 36.31857934, 28.81251646,
       54.72691223, 14.95149827, 80.41750109,  8.38051372, 98.70

In [3]:
# adding noise

noise= np.random.normal(loc=0, scale=0.5, size=100)  # loc--> whole data mean will be 0(mean=0) and scale(std deviation)=0.5
noise

# to make the data random we are adding noise

array([ 0.04352353, -0.14950368,  0.04588039, -0.99378446, -0.10983594,
        0.17855629,  0.73894702, -0.25913511, -0.4042468 , -0.25087852,
        0.45770106,  0.16437555, -0.2648801 ,  0.25663372,  0.04853877,
        0.4843225 , -0.35102655, -0.16383107, -0.19605408, -0.73175747,
        0.14806014,  0.13052764,  0.00255673, -0.11729357, -0.70768537,
       -0.21032266, -0.17135726, -0.40113863, -0.08064286,  0.20202543,
        0.94309295,  0.08728891,  0.1287752 , -0.03722296, -0.95938561,
       -0.01325694,  0.0301151 ,  1.23162106, -0.09618048,  0.15077367,
       -0.01735588, -0.58433902,  0.57141141,  0.37596652,  0.39551597,
       -0.45469373,  0.70139716, -0.70092553,  0.29342855,  1.09522781,
       -0.49526816, -0.28314886,  0.04982568, -0.25173783, -0.77533172,
        0.03428149, -0.53115186,  0.23679622, -0.45971212,  0.7749672 ,
       -0.39162665, -0.16103076,  0.40675861, -0.61543216,  0.11372997,
        0.65357138, -0.80374162,  0.09231693,  0.1299414 ,  0.39

In [4]:
y_data = ((X_data + noise) > 5).astype(int)

In [5]:
y_data

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1])

In [6]:
X_data

array([38.07947177, 95.12071633, 73.46740024, 60.26718994, 16.4458454 ,
       16.44345751,  6.7502776 , 86.75143843, 60.51038616, 71.0991852 ,
        3.03786494, 97.02107536, 83.41182144, 22.02157196, 19.00067175,
       19.15704648, 31.11998205, 52.95088673, 43.76255685, 29.83168488,
       61.57343658, 14.8098922 , 29.9223202 , 37.26982249, 46.15092844,
       78.73242018, 20.76770443, 51.9092094 , 59.64904232,  5.59859086,
       61.14694034, 17.88188825,  7.44010771, 94.93966819, 96.59757127,
       81.03133746, 31.15676315, 10.66953929, 68.73906962, 44.57509688,
       13.08178525, 50.0225141 ,  4.40446359, 91.02271981, 26.61921818,
       66.58970615, 31.85939653, 52.4867341 , 55.12431765, 19.3005911 ,
       96.98887815, 77.73814951, 94.01039521, 89.58790769, 60.1920979 ,
       92.26554927,  9.7607577 , 20.40230338,  5.4775016 , 33.20770275,
       39.47905168, 27.86355415, 83.04501341, 36.31857934, 28.81251646,
       54.72691223, 14.95149827, 80.41750109,  8.38051372, 98.70

In [7]:
# create a dataframe

data= {"feature":X_data, "target":y_data}
df= pd.DataFrame(data)
df

feature  target
0   38.079472       1
1   95.120716       1
2   73.467400       1
3   60.267190       1
4   16.445845       1
..        ...     ...
95  49.885764       1
96  52.750550       1
97  43.326561       1
98   3.516494       0
99  11.681251       1

[100 rows x 2 columns]

In [8]:
# convert df to pytorch tensor

X = torch.tensor(df['feature'].values, dtype= torch.float32).view(-1,1)
Y = torch.tensor(df['target'].values, dtype= torch.float32).view(-1,1)

In [9]:
X,Y

(tensor([[38.0795],
         [95.1207],
         [73.4674],
         [60.2672],
         [16.4458],
         [16.4435],
         [ 6.7503],
         [86.7514],
         [60.5104],
         [71.0992],
         [ 3.0379],
         [97.0211],
         [83.4118],
         [22.0216],
         [19.0007],
         [19.1570],
         [31.1200],
         [52.9509],
         [43.7626],
         [29.8317],
         [61.5734],
         [14.8099],
         [29.9223],
         [37.2698],
         [46.1509],
         [78.7324],
         [20.7677],
         [51.9092],
         [59.6490],
         [ 5.5986],
         [61.1469],
         [17.8819],
         [ 7.4401],
         [94.9397],
         [96.5976],
         [81.0313],
         [31.1568],
         [10.6695],
         [68.7391],
         [44.5751],
         [13.0818],
         [50.0225],
         [ 4.4045],
         [91.0227],
         [26.6192],
         [66.5897],
         [31.8594],
         [52.4867],
         [55.1243],
         [19.3006],


In [10]:
# we need to write everything inside class while using pytorch

class LogisticRegression(torch.nn.Module):
    def __init__(self, input_size,hidden_size1, hidden_size2, output_size):
        super(LogisticRegression, self).__init__()

        self.layer1 = nn.Linear(input_size, hidden_size1)  # layer 1
        self.relu1= nn.ReLU()

        self.layer2 = torch.nn.Linear(hidden_size1, hidden_size2)  # layer 2
        self.sigmoid1= nn.Sigmoid()

        self.output = torch.nn.Linear(hidden_size2, output_size)  # op layer

    def forward(self, x):
        x= self.layer1(x)
        x = self.relu1(x)
        x = self.layer2(x)
        x = self.sigmoid1(x)
        x = self.output(x)
        return x

In [11]:
input_size= 1  # number of featres
hidden_size1= 3
hidden_size2= 7
output_size= 1  # for binary classification- output size can be either 0 or 1

model= LogisticRegression(input_size, hidden_size1, hidden_size2, output_size)

In [12]:
criterion= nn.BCEWithLogitsLoss()  # loss
optimizer = optim.SGD(model.parameters(), lr=0.01)   # optimizer

In [13]:
num_epochs = 100
for epoch in range(num_epochs):
  # for forward popogation
  output= model(X)  # Y_predicted
  loss= criterion(output, Y)   # compare the output we calculated to Y(actual output)

  # for backward popogation
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()

  # print loss
  if (epoch+1) % 10 == 0:
    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

  # print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

Epoch [10/100], Loss: 0.6020
Epoch [20/100], Loss: 0.5146
Epoch [30/100], Loss: 0.4485
Epoch [40/100], Loss: 0.3979
Epoch [50/100], Loss: 0.3585
Epoch [60/100], Loss: 0.3273
Epoch [70/100], Loss: 0.3024
Epoch [80/100], Loss: 0.2820
Epoch [90/100], Loss: 0.2653
Epoch [100/100], Loss: 0.2514


In [14]:
# to see the matrix parameters

with torch.no_grad():
  print(f'Trained Parameter: layer1 weight = {model.layer1.weight.data}, layer1 bias ={model.layer1.bias.data}')
  print(f'Trained Parameter: layer2 weight = {model.layer2.weight.data}, layer2 bias ={model.layer2.bias.data}')


Trained Parameter: layer1 weight = tensor([[-0.1086],
        [-0.0553],
        [ 0.5728]]), layer1 bias =tensor([ 0.2370, -0.3988,  0.1477])
Trained Parameter: layer2 weight = tensor([[ 0.1747, -0.4654,  0.5538],
        [ 0.1557,  0.5624,  0.5739],
        [-0.5407, -0.4406,  0.5625],
        [-0.2627,  0.3335, -0.3042],
        [-0.5168,  0.2213, -0.4961],
        [ 0.5575, -0.2032,  0.3260],
        [ 0.0336,  0.0101,  0.4426]]), layer2 bias =tensor([ 0.0481,  0.4337, -0.0346,  0.4935,  0.2087,  0.5667,  0.2053])


In [15]:
with torch.no_grad():
  prediction = torch.sigmoid(model(X)).numpy()
  print(prediction)

[[0.8226102 ]
 [0.8225473 ]
 [0.82254744]
 [0.8225488 ]
 [0.8236709 ]
 [0.823671  ]
 [0.8130968 ]
 [0.8225473 ]
 [0.82254875]
 [0.82254755]
 [0.7873547 ]
 [0.8225473 ]
 [0.8225473 ]
 [0.8232589 ]
 [0.82352424]
 [0.8235113 ]
 [0.82274044]
 [0.82255256]
 [0.8225718 ]
 [0.82278323]
 [0.8225485 ]
 [0.82361674]
 [0.82277995]
 [0.8226192 ]
 [0.82256377]
 [0.8225474 ]
 [0.82336944]
 [0.8225535 ]
 [0.822549  ]
 [0.80767345]
 [0.82254857]
 [0.823608  ]
 [0.8155464 ]
 [0.8225473 ]
 [0.8225473 ]
 [0.8225474 ]
 [0.8227393 ]
 [0.82177126]
 [0.82254773]
 [0.82256866]
 [0.8232716 ]
 [0.82255584]
 [0.7998332 ]
 [0.8225473 ]
 [0.8229297 ]
 [0.82254785]
 [0.8227192 ]
 [0.8225529 ]
 [0.82255095]
 [0.8234991 ]
 [0.8225473 ]
 [0.8225474 ]
 [0.8225473 ]
 [0.8225473 ]
 [0.8225488 ]
 [0.8225473 ]
 [0.82067484]
 [0.82340217]
 [0.80699056]
 [0.8226861 ]
 [0.8225972 ]
 [0.82286555]
 [0.8225473 ]
 [0.8226313 ]
 [0.82282305]
 [0.8225512 ]
 [0.82362944]
 [0.8225474 ]
 [0.81812096]
 [0.8225473 ]
 [0.8225474 ]
 [0.82

In [16]:
df['prob'] = prediction
df['pred'] = (prediction > 0.5).astype(int)  # to convert the boolean values to int- True(0) False(1)
df['correct'] = (df['pred'] == df['target']).astype(int)

accuracy = np.mean(df['pred'] == df['target'])
accuracy

0.96

In [17]:
df.head()

feature  target      prob  pred  correct
0  38.079472       1  0.822610     1        1
1  95.120716       1  0.822547     1        1
2  73.467400       1  0.822547     1        1
3  60.267190       1  0.822549     1        1
4  16.445845       1  0.823671     1        1

In [18]:
# Pytorch With the drop-OUT ----
# we need to write everything inside class while using pytorch

class LogisticRegression(torch.nn.Module):
    def __init__(self, input_size,hidden_size1, hidden_size2, output_size):
        super(LogisticRegression, self).__init__()

        self.layer1 = nn.Linear(input_size, hidden_size1)  # layer 1
        self.relu1= nn.ReLU()
        self.dropout1 = nn.Dropout(0.5)

        self.layer2 = torch.nn.Linear(hidden_size1, hidden_size2)  # layer 2
        self.sigmoid1= nn.Sigmoid()
        self.dropout2 = nn.Dropout(0.5)

        self.output = torch.nn.Linear(hidden_size2, output_size)  # op layer

    def forward(self, x):
        x= self.layer1(x)
        x = self.relu1(x)
        x = self.dropout1(x)
        x = self.layer2(x)
        x = self.sigmoid1(x)
        x = self.dropout2(x)
        x = self.output(x)
        return x

In [20]:
input_size= 1  # number of features
hidden_size1= 3
hidden_size2= 7
output_size= 1  # for binary classification- output size can be either 0 or 1

model= LogisticRegression(input_size, hidden_size1, hidden_size2, output_size)

In [21]:
criterion= nn.BCEWithLogitsLoss()  # loss
optimizer = optim.SGD(model.parameters(), lr=0.01)   # optimizer

In [22]:
num_epochs = 100
for epoch in range(num_epochs):
  # for forward popogation
  output= model(X)  # Y_predicted
  loss= criterion(output, Y)   # compare the output we calculated to Y(actual output)

  # for backward popogation
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()

  # print loss
  if (epoch+1) % 10 == 0:
    print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

Epoch [10/100], Loss: 0.4805
Epoch [20/100], Loss: 0.4529
Epoch [30/100], Loss: 0.4261
Epoch [40/100], Loss: 0.4356
Epoch [50/100], Loss: 0.4090
Epoch [60/100], Loss: 0.3833
Epoch [70/100], Loss: 0.3730
Epoch [80/100], Loss: 0.3451
Epoch [90/100], Loss: 0.3511
Epoch [100/100], Loss: 0.2966


In the last epoch, the loss is 0.2966, but we have acheived lower Loss without the use of drop out, so in this case we can ignore using it.

- Drop-out is also hyperparameter


In [23]:
# to see the matrix parameters

with torch.no_grad():
  print(f'Trained Parameter: layer1 weight = {model.layer1.weight.data}, layer1 bias ={model.layer1.bias.data}')
  print(f'Trained Parameter: layer2 weight = {model.layer2.weight.data}, layer2 bias ={model.layer2.bias.data}')


Trained Parameter: layer1 weight = tensor([[ 0.3153],
        [-0.0899],
        [-0.7732]]), layer1 bias =tensor([ 0.1659,  0.5518, -0.2062])
Trained Parameter: layer2 weight = tensor([[-1.1104e-01,  3.9611e-01,  4.2824e-02],
        [-2.8093e-01,  1.8958e-01, -5.2554e-01],
        [-4.7998e-01,  2.1831e-01,  1.0653e-01],
        [-2.8543e-01, -5.4753e-01,  3.0876e-01],
        [ 1.4111e-01, -5.7500e-01, -4.6624e-01],
        [ 2.9096e-01, -3.7191e-01,  5.3157e-02],
        [ 3.9742e-01, -2.4648e-04,  2.3751e-01]]), layer2 bias =tensor([ 0.0701, -0.3969, -0.0054, -0.0689,  0.0812, -0.5509,  0.4038])


In [24]:
df['prob'] = prediction
df['pred'] = (prediction > 0.5).astype(int)  # to convert the boolean values to int- True(0) False(1)
df['correct'] = (df['pred'] == df['target']).astype(int)

accuracy = np.mean(df['pred'] == df['target'])
accuracy

0.96